### DISCUSSION TITLE
top-level command list

### DISCUSSION THREAD

**Discussion by isaacs** [vltpkg/vltpkg#63] [Ideas]
In an effort to minimize the number of nouns and verbs that users need to know about, while balancing ergonomic affordances (especially those that are appreciated by yarn/pnpm/npm users) and making any potentially insecure actions very explicit, we should try to be thoughtful about which top-level commands we define.

This is an exploration, please weigh in with your thoughts!

## do

- install (i, add)
- uninstall (u, rm)
- run (r, vlr) - run a script from package.json
- exec (x, vlx) - run a bin defined by a dependency (ie, `exec(argv)` but with `node_modules/.bin` in the PATH), no auto-install
- install-exec (ix, vlix) - run a bin defined by a dep if present, installing if necessary (like `npx`)
- run-exec (rx, vlrx) - run if the script exists, otherwise exec. (like pnpm's fallback)
- query - do fancy stuff with dependency selector syntax (maybe could fill in for stuff like `npm fund`, etc?)
- config (conf) - manage config settings (get, set, edit, list, etc.)
- workspace (ws) - manage workspaces
- update - do we need this? Would it be better as a flag to `vlt i`?
- help (h, ?)
- (name tbd) something to manage logins and auth on various registries? (satisfies login, token, profile, org)
- publish (p)
- (name tbd) something to manage various registry metadata (eg, `deprecate`, `dist-tag`, etc.)
- list (ls) - maybe not needed, if we have query?
- audit/audit fix - maybe we instead have a plugin or something that talks to a security provider or CVE source?

## do not

- test, start, stop - use `run` or `vlr`
- fund - use `vlt query`
- outdated - could `vlt query` do this? (if not, maybe we need this?)
- view - could `vlt query` do this? (if not, maybe we need this?)
- explain - use `vlt query`
- explore, edit - not sure anyone uses these much (needs research, perhaps via the "ship without the feature and see how many complaints we get" method)
- ci - let's just do the fast thing every time by default
- prune - let's just not make garbage in node_modules, prune up front on install

**Comment by ruyadorno** [comment]
I believe we should have a shorter list for an MVP, focusing only in the very minimal necessary to install packages:

- install (i, add)
- uninstall (u, rm)
- help (h, ?)
- login / auth / tbd
- publish (p)

**Reply by isaacs** [reply]
I was thinking the same thing initially, but:

- `config` is eventually important, and a good first command to flesh out the CLI architecture, since it doesn't depend on anything _other_ than configuration, but is still complicated enough to flesh out the patterns. So, even though we probably don't _need_ need it first, it's a good dev spike, and we may as well ship it.
- We have to do most of the work for `run` in order to do `install` anyway, since it's almost the same code paths for "run this project's scripts" and "run this dep's scripts".
- If we have `run`, then `exec` is trivial. It's just "exec this command" instead of "look up this command by name and then exec it".
- If we have `run` and `exec` and `install`, then `run-exec` and `install-exec` are trivial.
- For an MVP we'll need at least _some_ kind of reasonable output for `vlt -h`, and `help` is an easy one to iterate on, so this is also a sort of "dev spike" in the same vein as `config`.

I agree about the others, though:

- `update` will need to implement almost immediately after #14 lands, but definitely lower priority, and probably easy to do once we have `install`.
- `workspace` can be managed manually for now, maybe even put off until after the initial release, since it's just DX affordances around a fairly simple config/fs interface.
- `query` is potentially a huge feature, especially if we want to make it as powerful as it'll need to be, so definitely punt on it for now.

Of course, yagni principle is important, avoid scope creep, etc, but also, it's important to not do extra work to _avoid_ doing something we know we'll need to do eventually.

Imo, the Big Rocks with unknown-unknowns are:

- install
- login
- publish

**Reply by ruyadorno** [reply]
yes, that all makes sense! 

Now one thing that comes to my mind is that I prefer having the "update" functionality turned into some sort of option/flag of the `install` command though instead of having it be a top-level command.

**Reply by isaacs** [reply]
Better to just make it a top-level command, imo. The surface area is actually the same either way, because we're still exposing the same amount of functionality, but it's more explicit and makes the code clearer to expose with a standalone command.

For starters, I'm not sure that `vlt install --update` is any easier for users to remember/guess than `vlt update`. In fact, it's almost certainly the opposite, and likely makes the code more confusing, because now it has to use different strategies based on the config flag.

For example, what does `vlt prune --update` do? Or `vlt install-exec --update`? Or even `vlt install thing --update`, would that install `thing`, but then update everything else?

If it's a configuration flag, we have to consider its impact _anywhere_ we might be doing something with the installed graph on disk. If it's a top-level command, it's isolated and very clear that you only get update behavior when you're running the update command.

**Comment by ruyadorno** [comment]
### audit

I would like to hear you all thoughts on audit.

Personally I don't believe we should ever audit by default on install (like npm does) and I believe the `audit` command should be on the **do not** list or at most be the very last item in the **do** list.

**Reply by ruyadorno** [reply]
I do like the idea of potentially partner with a security provider for that.

**Reply by isaacs** [reply]
The only reason I put it on the *do* list is that we've seen people excitedly assuming that what we're doing is "building a more secure package manager". Not that that's _not_ what we're doing, of course, but we're focused more on speed and basic functionality for now (which I think is the right approach), and it feels like we can't really call it ready for prime time until there's some story there. Maybe not npm's fairly naive "check for CVE hits and warn", but we'll need something soon. A partnership with socket, snyk, garnet.ai (or someone else, or all of these with some kind of plugin thingie) makes sense here, I think.

**Reply by isaacs** [reply]
Another thought: since a lot of the concern with "auto-audit on install" is performance, I wonder if we could maybe save the perf hit by doing audit out of band. Ie, on install, we kick off a daemon that'll hit the security endpoint, and write the results to a known filename. Then on the _next_ install, we'd load those results and use it to report issues. So it'd seem instantaneous, albeit only on the second install.

**Comment by ruyadorno** [comment]
### run-exec

Alright, please educate me on this one, what are the usecases? do we want to do it just to have a drop-in replacement to `pnpm` fallback command style? is there more into it?

My initial (and uninformed) gut feeling is that maybe this shouldn't be part of the initial set of commands at all 🤔

**Reply by isaacs** [reply]
Yes, it is a drop-in replacement for pnpm's default behavior. But also, let's take some time to appreciate just how great that default behavior is, because of the mindset that it suggests about the tool being a "portal" to all things about your dependency graph.

If you want to run a program that's provided by your dependencies, you don't have to think at all about whether it's a bin that's provided by a dependency, or a package.json script that references them. It's pretty common to see stuff in npm-land like:

```json
{
  "scripts": {
    "tap": "tap",
    "eslint": "eslint",
    "prettier": "prettier"
  }
}
```

just so that people can do `npm run tap ...`, because (reasonably or not! we are _very_ lazy creatures) that way you don't have to remember that "these things I do with `npm run`, these other things I do with `npm exec`. But you still need to assign this scripts alias to register everything you'll get access to, which has been reported as a pain point multiple times.

With pnpm, you have this one command, just `pnpm`, which is your doorway to all your dependency programs. When you run `pnpm whatever`, you're saying that you want whatever `whatever` is that lives in your dependencies, your package.json, etc. Pnpm knows about it, so it'll give it to you.

Now, I _don't_ think it should be the fallback by default. It should be opt-in. And there are cases where it's _good_ to be very explicit, like "no, don't just run the `eslint` in node_modules, run _this specific script_, with pre/post/etc".

With `run-exec`, you do also get an interesting capability from the ambiguity. You can use `pnpm eslint` today, and then if we decide, "you know what? In this one project, we should run it with this extra flag", we can just add the alias in package.json#scripts, and no one has to change their muscle memory. `pnpm eslint` will just run the script if it exists, _otherwise_ it'll run the bin that it finds in `nm/.bin`. You can also set up pre/post hooks on it, etc. But for lighter weight use cases, there's no need, so you save the extra step.

Regardless of the arguments for or against, though, the nail-in-the-coffin for me is that _people do it_. I see in issues almost daily in tap and other projects, people run their tests by doing `pnpm tap`, build code with `pnpm tsc`, etc, sometimes without even bothering to have a `"scripts": {"test"}` set up at all. When they do set up a script, they start running `yarn build` instead of `yarn run build`. (Even as a long-time npm user, how often do you type `npm run-script ...` instead of `npm run`? We are all very lazy.)

As a new tool in the space, everything we can do to go with the flow of existing user behavior is an opportunity. There are enough unavoidable differences that we'll be asking users to accept; the smaller we can keep the set of required behavioral changes, the more successful vlt will be.

While we should definitely not take on unnecessary (or especially, _uncertain_) work at this stage, there's little benefit to keeping the initial set of commands artificially or inconveniently small. For a correct install, we _must_ have the code that drives `vlt run` and `vlt exec`, and they're important parts of the dev tool in their own right, that we likely cannot succeed without. So why not just wire them in? It's not much work, and we know it's not wasted. In similar fashion, if we have `vlt run` and `vlt exec`, then `vlt run-exec` is trivial.

If we have to do 90% of the work anyway, finishing the job along the way is less work overall than having to come back later, when the context is not as fresh in our minds. (It's a small thing either way, but why put it off? It's never going to be easier than it is right now.)

By contrast, I am very confident that we'll need `vlt query` (or something like it) eventually. If we ship a 1.0 without a way to inspect and interact with the dep graph, we're going to need to do it immediately. However, no part of it is on the path to an end-to-end install, so implementing that now would be a distraction, and a significant amount of work.

(For me personally, I probably would set the fallback command to `run-exec`, but also, I have a bash shortcut that puts `./node_modules/.bin;../node_modules/.bin;...` in the `PATH` env, so locally installed stuff is always available. That's way less friction day to day, but a pretty considerable up-front investment that most people will never bother to do.)
